### PR TITLE
Dev/amk rf predict

### PR DESCRIPTION
**Pull request by makart19**
Added RF (classification & regression) predictor for GPU

### CODE REVIEW

File: cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/cl_kernels/df_batch_predict_classification_kernels.cl [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+        iEnd = (iEnd > nRows) ? nRows : iEnd;
+
+        if (tree_id < nTrees)
+        {
+            const __global int * ftrIdxForTree                   = ftrIdx + tree_id * maxTreeSize;
+            const __global int * classLabelsOrNextNodeIdxForTree = classLabelsOrNextNodeIdx + tree_id * maxTreeSize;
+            const __global algorithmFPType * ftrValueForTree     = ftrValue + tree_id * maxTreeSize;
+            const __global double * classProbaForTree            = classProba + tree_id * maxTreeSize * nClasses;
+
+            uint treeRootIsSplit = (uint)(-1 != ftrIdxForTree[0]);
```
Comment (michael-smirnov) [inline cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/cl_kernels/df_batch_predict_classification_kernels.cl:60]: what `ftr` stands for? `-1` - magic number

File: cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/df_classification_predict_dense_oneapi_impl.i [hunk 2]
```diff
@@ -0,0 +1,495 @@
...
+{
+namespace decision_forest
+{
+namespace classification
+{
+namespace prediction
+{
+namespace internal
+{
+inline char * utoa(uint32_t value, char * buffer, size_t buffer_size)
```
Comment (michael-smirnov) [inline cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/df_classification_predict_dense_oneapi_impl.i:60]: Please check with @averbukh that this function satisfies SDL. I saw there are restriction about using null-terminated strings

File: cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/df_classification_predict_dense_oneapi_impl.i [hunk 3]
```diff
@@ -0,0 +1,495 @@
...
+{
+    services::Status status;
+
+    _nClasses     = nClasses;
+    _votingMethod = votingMethod;
+    const daal::algorithms::decision_forest::classification::internal::ModelImpl * const pModel =
+        static_cast<const daal::algorithms::decision_forest::classification::internal::ModelImpl * const>(m);
+
+    services::String buildOptions = getBuildOptions(_nClasses);
+    //DAAL_CHECK_STATUS_VAR(_treeLevelBuildHelper.init(buildOptions.c_str()));
```
Comment (michael-smirnov) [inline cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/df_classification_predict_dense_oneapi_impl.i:136]: delete

File: cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/df_classification_predict_dense_oneapi_impl.i [hunk 4]
```diff
@@ -0,0 +1,495 @@
...
+
+    UniversalBuffer classHist;
+    if (prob)
+    {
+        DAAL_CHECK_STATUS_VAR(const_cast<NumericTable *>(prob)->getBlockOfRows(0, nRows, readWrite, probBlock));
+        classHist = probBlock.getBuffer();
+    }
+    else
+    {
+        classHist = context.allocate(TypeIds::id<algorithmFPType>(), _nClasses * nRows, &status);
```
Comment (michael-smirnov) [inline cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/df_classification_predict_dense_oneapi_impl.i:171]: do we need an overflow check here?

File: cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/df_classification_predict_dense_oneapi_impl.i [hunk 5]
```diff
@@ -0,0 +1,495 @@
...
+    _nTreeGroups = 8;
+
+    if (nTrees > 48)
+    {
+        _nTreeGroups = 32;
+    }
+    else if (nTrees > 12)
+    {
+        _nTreeGroups = 16;
+    }
```
Comment (michael-smirnov) [inline cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/df_classification_predict_dense_oneapi_impl.i:215]: A lot of magic numbers

File: cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/df_classification_predict_dense_oneapi_impl.i [hunk 6]
```diff
@@ -0,0 +1,495 @@
...
+        {
+            KernelArguments args(12);
+            args.set(0, srcBuffer, AccessModeIds::read);
+            args.set(1, featureIndexList, AccessModeIds::read);
+            args.set(2, leftOrClassTypeList, AccessModeIds::read);
+            args.set(3, featureValueList, AccessModeIds::read);
+            args.set(4, classProba, AccessModeIds::read);
+            args.set(5, obsClassHist, AccessModeIds::readwrite);
+            args.set(6, scale);
+            args.set(7, nRows);
```
Comment (michael-smirnov) [inline cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/df_classification_predict_dense_oneapi_impl.i:342]: Kernel accepts `int` but not `size_t`

File: cpp/daal/src/algorithms/dtrees/forest/regression/oneapi/cl_kernels/df_batch_predict_regression_kernels.cl [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+        const int local_size    = get_local_size(0);
+        const int n_groups      = get_num_groups(0);
+        const int group_id      = get_group_id(0);
+        const int n_tree_groups = get_num_groups(1);
+        const int tree_group_id = get_group_id(1);
+        const int tree_id       = treeOffset + tree_group_id;
+
+        const int nElementsForGroup = nRows / n_groups + !!(nRows % n_groups);
+
+        int iStart = group_id * nElementsForGroup;
```
Comment (michael-smirnov) [inline cpp/daal/src/algorithms/dtrees/forest/regression/oneapi/cl_kernels/df_batch_predict_regression_kernels.cl:46]:
```suggestion
        const int iStart = group_id * nElementsForGroup;
```

File: cpp/daal/src/algorithms/dtrees/forest/regression/oneapi/df_regression_predict_dense_kernel_oneapi.h [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+#include "data_management/data/numeric_table.h"
+#include "algorithms/algorithm_base_common.h"
+#include "algorithms/decision_forest/decision_forest_regression_predict.h"
+#include "src/algorithms/dtrees/forest/regression/df_regression_model_impl.h"
+#include "algorithms/decision_forest/decision_forest_regression_model.h"
+
+using namespace daal::data_management;
+using namespace daal::services;
+using namespace daal::oneapi::internal;
+using namespace daal::algorithms::dtrees::internal;
```
Comment (michael-smirnov) [inline cpp/daal/src/algorithms/dtrees/forest/regression/oneapi/df_regression_predict_dense_kernel_oneapi.h:39]: do not do `using namespace` in headers

**Review by KulikovNikita — APPROVED** [review]

File: cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/cl_kernels/df_batch_predict_classification_kernels.cl (hunk 1 above)
Comment (makart19) [reply cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/cl_kernels/df_batch_predict_classification_kernels.cl]: leaf node, fixed

File: cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/df_classification_predict_dense_oneapi_impl.i (hunk 2 above)
Comment (makart19) [reply cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/df_classification_predict_dense_oneapi_impl.i]: thanks, migrated to daal_int_to_string

File: cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/df_classification_predict_dense_oneapi_impl.i (hunk 3 above)
Comment (makart19) [reply cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/df_classification_predict_dense_oneapi_impl.i]: fixed

File: cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/df_classification_predict_dense_oneapi_impl.i (hunk 4 above)
Comment (makart19) [reply cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/df_classification_predict_dense_oneapi_impl.i:171]: added, thanks

File: cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/df_classification_predict_dense_oneapi_impl.i (hunk 5 above)
Comment (makart19) [reply cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/df_classification_predict_dense_oneapi_impl.i]: fixed

File: cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/df_classification_predict_dense_oneapi_impl.i (hunk 6 above)
Comment (makart19) [reply cpp/daal/src/algorithms/dtrees/forest/classification/oneapi/df_classification_predict_dense_oneapi_impl.i]: fixed

File: cpp/daal/src/algorithms/dtrees/forest/regression/oneapi/df_regression_predict_dense_kernel_oneapi.h (hunk 8 above)
Comment (makart19) [reply cpp/daal/src/algorithms/dtrees/forest/regression/oneapi/df_regression_predict_dense_kernel_oneapi.h]: fixed, thanks

File: cpp/daal/src/algorithms/dtrees/forest/regression/oneapi/cl_kernels/df_batch_predict_regression_kernels.cl (hunk 7 above)
Comment (makart19) [reply cpp/daal/src/algorithms/dtrees/forest/regression/oneapi/cl_kernels/df_batch_predict_regression_kernels.cl]: fixed

File: cpp/daal/src/algorithms/dtrees/forest/regression/oneapi/df_regression_predict_dense_kernel_oneapi.h [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+    const size_t _nRowsBlocksForMedium = 8;
+
+    const size_t _nTreesLarge  = 192;
+    const size_t _nTreesMedium = 48;
+    const size_t _nTreesSmall  = 12;
+
+    const size_t _nTreeGroupsForLarge  = 128;
+    const size_t _nTreeGroupsForMedium = 32;
+    const size_t _nTreeGroupsForSmall  = 16;
+    const size_t _nTreeGroupsMin       = 8;
```
Comment (michael-smirnov) [inline cpp/daal/src/algorithms/dtrees/forest/regression/oneapi/df_regression_predict_dense_kernel_oneapi.h:89]: It's worth briefly comment why these numbers were chosen for different workload sizes.

**Review by michael-smirnov — APPROVED** [review]
Minor comments

File: cpp/daal/src/algorithms/dtrees/forest/regression/oneapi/df_regression_train_hist_kernel_oneapi.h [hunk 10]
```diff
@@ -35,7 +35,7 @@
 #include "src/algorithms/dtrees/forest/oneapi/df_feature_type_helper_oneapi.h"
 #include "src/algorithms/dtrees/forest/oneapi/df_tree_level_build_helper_oneapi.h"
 
-using namespace daal::data_management;
+//using namespace daal::data_management;
```
[the remark this answers is not in the document]
Comment (makart19) [reply cpp/daal/src/algorithms/dtrees/forest/regression/oneapi/df_regression_train_hist_kernel_oneapi.h]: fixed

File: cpp/daal/src/algorithms/dtrees/forest/regression/oneapi/df_regression_predict_dense_kernel_oneapi.h (hunk 9 above)
Comment (makart19) [reply cpp/daal/src/algorithms/dtrees/forest/regression/oneapi/df_regression_predict_dense_kernel_oneapi.h:90]: added comment

**Comment by makart19** [comment]
@Mergifyio backport rls/2021-gold-mnt

**Pull request merged**